File: frontend/android/app/src/main/java/ch/duartesantos/opengym/PrintPlugin.java
package ch.duartesantos.opengym;

import android.content.Context;
import android.print.PrintAttributes;
import android.print.PrintDocumentAdapter;
import android.print.PrintManager;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * Minimal local Capacitor plugin that hands a self-contained HTML document to the Android
 * system print flow (Save as PDF, Save to Drive, a real printer). Android WebView has no
 * window.print(), so the printable plan page (lib/plan-share.js planPrintHTML) can only reach
 * a PDF this way. No PDF library is bundled — the OS renders it.
 *
 * Usage from JS:
 *   import { registerPlugin } from '@capacitor/core';
 *   const Print = registerPlugin('Print');
 *   await Print.printHtml({ html: '<!doctype html>…', name: 'Weekly Training Plan' });
 */
@CapacitorPlugin(name = "Print")
public class PrintPlugin extends Plugin {

    // The print framework drives an off-screen WebView asynchronously; without a strong
    // reference it can be collected mid-spool and the job dies silently. Held until the next
    // print replaces it — one at a time is all the UI can start.
    private WebView printView;

    @PluginMethod
    public void printHtml(PluginCall call) {
        String html = call.getString("html");
        if (html == null || html.isEmpty()) {
            call.reject("html is required");
            return;
        }
        final String jobName = call.getString("name", "openGym");

        getActivity().runOnUiThread(() -> {
            WebView webView = new WebView(getContext());
            webView.setWebViewClient(new WebViewClient() {
                @Override
                public void onPageFinished(WebView view, String url) {
                    PrintManager printManager =
                            (PrintManager) getContext().getSystemService(Context.PRINT_SERVICE);
                    if (printManager == null) {
                        printView = null;
                        call.reject("printing is not available on this device");
                        return;
                    }
                    PrintDocumentAdapter adapter = view.createPrintDocumentAdapter(jobName);
                    printManager.print(jobName, adapter,
                            new PrintAttributes.Builder()
                                    .setMediaSize(PrintAttributes.MediaSize.ISO_A4)
                                    .build());
                    call.resolve();
                }
            });
            // baseURL null: the document is fully self-contained (inline <style>, no assets).
            webView.loadDataWithBaseURL(null, html, "text/html", "UTF-8", null);
            printView = webView;
        });
    }
}

File: frontend/ios/App/App/PrintPlugin.swift
import Foundation
import Capacitor
import UIKit

/**
 * Minimal local Capacitor plugin that hands a self-contained HTML document to the iOS system
 * print sheet (Save to Files as PDF, share, a real printer). The printable plan page
 * (lib/plan-share.js planPrintHTML) reaches a PDF this way without bundling a PDF library —
 * UIKit renders it.
 *
 * Usage from JS:
 *   import { registerPlugin } from '@capacitor/core';
 *   const Print = registerPlugin('Print');
 *   await Print.printHtml({ html: '<!doctype html>…', name: 'Weekly Training Plan' });
 */
@objc(PrintPlugin)
public class PrintPlugin: CAPPlugin {

    @objc func printHtml(_ call: CAPPluginCall) {
        guard let html = call.getString("html"), !html.isEmpty else {
            call.reject("html is required")
            return
        }
        let jobName = call.getString("name") ?? "openGym"

        DispatchQueue.main.async {
            let controller = UIPrintInteractionController.shared

            let info = UIPrintInfo(dictionary: nil)
            info.outputType = .general
            info.jobName = jobName
            controller.printInfo = info

            // The document carries its own @page margins; keep the formatter's own insets at 0
            // so they don't stack on top.
            let formatter = UIMarkupTextPrintFormatter(markupText: html)
            formatter.perPageContentInsets = .zero
            controller.printFormatter = formatter

            controller.present(animated: true) { (_, completed, error) in
                if let error = error {
                    call.reject(error.localizedDescription)
                } else {
                    call.resolve(["completed": completed])
                }
            }
        }
    }
}
